### ISSUE TITLE
Tolerance issue in Phase I of Chebfun2 constructor

### ISSUE THREAD

**Issue by bhashemi** [chebfun/chebfun#1491]
Consider the following example:
```
>> f = chebfun2(@(x,y) sin(80*x+y));
>> length(f.rows(:,1))
ans =
   257
>> length(f.cols(:,1))
ans =
   129
```
Having 257 coefficients in the x-direction is reasonable but it seems that we should be able to reduce the number of coefficients in the y-direction from 129 to 33 or at most 65. The reason we have this issue is that the grids used in Phase I of the Chebfun2 constructor are always "square" 
```
[xx, yy] = points2D(no_of_points, no_of_points, dom, pref);
```
and we allow rectangular grids in Phase II. We might want to improve Chebfun2 but this needs a change in the constructor. @ajt60gaibb, would you please let me know if I'm missing a point here?

**Comment by ajt60gaibb** [comment]
I am not sure what is happening here: `sin(80x+y)` is a rank 4 function. It should go to phase 2 on a grid of size 17 or maybe 33.

Sounds like a tolerance issue to me.

**Comment by bhashemi** [comment]
Yes I agree. This is a tolerance issue. We can see a rapid decay in the values of the first two pivots in comparison with the remaining pivots. 

![untitled](https://cloud.githubusercontent.com/assets/9217439/7610631/4773a2ec-f976-11e4-9bcf-6a34ffe6048c.png)

**Comment by ajt60gaibb** [comment]
So it looks like the tolerance needs to depend on derivative information. 

This makes sense.  See: http://en.wikipedia.org/wiki/Condition_number#Several_variables 
(It should probably be respecting the absolute condition number of a function, not that relative one).

**Comment by bhashemi** [comment]
Here is my suggestion for a new `tol` in Chebfun2 constructor, based on the above suggestion by @ajt60gaibb :
`
df_dx = diff(vals,1,2) ./ diff(xx,1,2); % xx changes column-wise

df_dy = diff(vals,1,1) ./ diff(yy,1,1); % yy changes row-wise

df_dx = df_dx (1:size(df_dx,2),:); % Throw out last row of df_dx, where we don't have df_dy.

df_dy = df_dy (:, 1:size(df_dy,1),:); % Throw out last column of df_dy, where we don't have df_dx.

J = max(abs(df_dx),abs(df_dy));

Jac_norm = max(J(:)) % an approximation for the norm of the Jacobian over the whole domain. 

tol = grid.^(2/3) * max( max( abs(dom(:))), 1) * Jac_norm * pseudoLevel
`

As you see I just replaced `vscale` in the formula for `tol` by `Jac_norm` which is somewhat the main ingredient of the absolute condition number of the 2D function evaluation. `Jac_norm` also changes as `vscale` does. Any ideas or suggestions as I might have computed the `Jac_norm` incorrectly?

**Comment by ajt60gaibb** [comment]
Seems reasonable to me.   Not sure about the absolute maximum entry of the Jacobian. The infinite norm makes more sense, perhaps? Or maybe not? It's a factor of at most `2` anyway... 

Does it work?

**Comment by bhashemi** [comment]
Yes, it works for `f = chebfun2(@(x,y) sin(80*x+y))`. The rank is now 2 (instead of the old value of 7) and we have:
`
>> length(f.rows(:,1))
ans =
   257
>> length(f.cols(:,1))
ans =
    33
>> 
`

I didn't try other functions though.

**Comment by bhashemi** [comment]
I agree that the inf-norm makes more sense. So, I replaced the max-entry with the inf-norm. The above result remains the same. Here is how it looks as a function:
`
function tol = newTol(xx, yy, vals, grid, dom, pseudoLevel) 
`
`
df_dx = diff(vals,1,2) ./ diff(xx,1,2); % xx changes column-wise
`
`
df_dy = diff(vals,1,1) ./ diff(yy,1,1); % yy changes row-wise
`
`
df_dx = df_dx (1:size(df_dx,2),:); % Throw out last row of df_dx, where we don't have df_dy.
`
`
df_dy = df_dy (:, 1:size(df_dy,1),:); % Throw out last column of df_dy, where we don't have df_dx.
`
`
J = max(abs(df_dx),abs(df_dy));
`
`
Jac_norm = norm(J, inf) % an approximation for the norm of the Jacobian over the whole domain.
`
`
tol = grid.^(2/3) * max( max( abs(dom(:))), 1) * Jac_norm * pseudoLevel
`
`
end
`

**Comment by ajt60gaibb** [comment]
@bhashemi, are you sure it should not be  `max_{x,y} ( \|J( x,y )\|_\infty )`?  It does not look like your code is doing that... or maybe I am mistaken? 

Anyway, if all the tests pass then something like this must be an improvement.  Great stuff. 

Does this tolerance also go into phase 2?  I guess it should do.

**Comment by bhashemi** [comment]
Let's review what we are supposed to do here. The Jacobain of the 2D function `f` is a row vector containing two partial derivatives at a specific point (x,y). We compute its `inf-norm` as follows
`
J = max(abs(df_dx),abs(df_dy));
`
So, this `J` is a _matrix_ containing inf-norm of the Jacobians at all the points. Then, we should have computed the maximum of `J` over all x and y.  This is what 
`
Jac_norm = max(J(:))
`
computes. Is this what I should do? 
By the way, 6 tests fail in `chebtest('chebfun2')` with this which I'll try to check further.

**Comment by bhashemi** [comment]
I had to apply 2 modifications to the formula in order to pass all the tests in `chebtest('chebfun2')`. Here are the two changes:
1) For the function `f = @(x,y) (1./(1+1e-3*(((x+1)/2-.3).^2+((y+1)/2-.3).^2)));` which is the last function in `chebfun2/test_battery.m`, we have 
`Jac_norm = 6.9452e-04`. Therefore, the formula
`
tol = grid.^(2/3) * max( max( abs(dom(:))), 1) * Jac_norm * pseudoLevel
`
gives us `tol = 1.0196e-18`. This causes the above test to pass after 200 seconds after giving a warning that the function is not low-rank.  Since we have `vscale = 1.0000` and the idea was to make the `tol` more reasonable by giving it information about the function derivative, I suggest to use the following formula:
`
tol = grid.^(2/3) * max( max( abs(dom(:))), 1) * max( Jac_norm, vscale) * pseudoLevel;
`
This now brings `tol = 1.4680e-15` and the test passes early.

2) The second test in `chebfun2/test_scl.m` is about horizontal invariance for the following function:
`
g = chebfun2( @(x,y) cos(x/eps.*y/eps), eps*[-1 1 -1 1] );
`
which fails because the formula 
`
tol = grid.^(2/3) * max( max( abs(dom(:))), 1) * max( Jac_norm, vscale) * pseudoLevel;
`
gives us `tol = 5.5287`. The reason here is the term `max( max( abs(dom(:))), 1)` in the formula. Once we simplify it to `max( abs(dom(:) ) )`, we get `tol =   1.2276e-15` and this test passes. The formula takes care of the scale of the domain when computing `df_dx` and `df_dy`.

To conclude, all the tests pass using the formulae
`
Jac_norm = max(J(:));
`
`
tol = grid.^(2/3) * max( abs(dom(:) ) ) * max( Jac_norm, vscale) * pseudoLevel;
`
and the example `f = chebfun2(@(x,y) sin(80*x+y));` we started this discussion from is also in good shape. Any ideas of any possible contradictions here?
